### PR TITLE
[SPARK-6526][ML] Add Normalizer transformer in ML package

### PR DESCRIPTION
**Pull request by yinxusen**
See [SPARK-6526](https://issues.apache.org/jira/browse/SPARK-6526).

@mengxr Should we add test suite for this transformer? There is no test suite for all feature transformers in ML package now.

### CODE REVIEW

**Comment by jkbradley** [comment]
We should definitely add a test suite for each transformer, especially as the package becomes less Experimental.

File: mllib/src/main/scala/org/apache/spark/ml/feature/Normalizer.scala [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+ * Normalize a vector into another one with a given p-norm.
+ */
+@AlphaComponent
+class Normalizer extends UnaryTransformer[Vector, Vector, Normalizer] {
+
+  /**
+   * p norm of the vector
+   * @group param
+   */
+  val p = new DoubleParam(this, "p", "p-norm", Some(2))
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Normalizer.scala]: The built-in param doc should probably say more than just "p-norm"
In the Scaladoc, we should mention the default value

File: mllib/src/main/scala/org/apache/spark/ml/feature/Normalizer.scala [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+import org.apache.spark.annotation.AlphaComponent
+import org.apache.spark.ml.UnaryTransformer
+import org.apache.spark.ml.param.{DoubleParam, ParamMap}
+import org.apache.spark.mllib.feature
+import org.apache.spark.mllib.linalg.{VectorUDT, Vector}
+import org.apache.spark.sql.types.DataType
+
+/**
+ * :: AlphaComponent ::
+ * Normalize a vector into another one with a given p-norm.
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Normalizer.scala]: "Normalize a vector to have unit norm using the given p-norm"

**Comment by yinxusen** [comment]
@jkbradley Yes, I think so. I'll add test suite soon.

File: mllib/src/test/scala/org/apache/spark/ml/feature/NormalizerSuite.scala [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+package org.apache.spark.ml.feature
+
+import org.scalatest.FunSuite
+
+import org.apache.spark.mllib.linalg.{DenseVector, SparseVector, Vector, Vectors}
+import org.apache.spark.mllib.util.MLlibTestSparkContext
+import org.apache.spark.mllib.util.TestingUtils._
+import org.apache.spark.sql.{DataFrame, Row, SQLContext}
+
+case class DataSet(features: Vector)
```
Comment (jkbradley) [inline mllib/src/test/scala/org/apache/spark/ml/feature/NormalizerSuite.scala]: Make this private, just as good practice.  Or better yet, you can call toDF("features") if you import sqlContext.implicits._ and not use a case class at all.

File: mllib/src/test/scala/org/apache/spark/ml/feature/NormalizerSuite.scala [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    }, "The vector type should be preserved after normalization.")
+
+    assert(result(0) ~== Vectors.sparse(3, Seq((0, -0.65617871), (1, 0.75460552))) absTol 1E-5)
+    assert(result(1) ~== Vectors.dense(0.0, 0.0, 0.0) absTol 1E-5)
+    assert(result(2) ~== Vectors.dense(0.184549876, -0.3383414, -0.922749378) absTol 1E-5)
+    assert(result(3) ~== Vectors.sparse(3, Seq((1, 0.27352993), (2, 0.96186349))) absTol 1E-5)
+    assert(result(4) ~== Vectors.dense(0.897906166, 0.113419726, 0.42532397) absTol 1E-5)
+    assert(result(5) ~== Vectors.sparse(3, Seq()) absTol 1E-5)
+  }
+  test("Normalization with setter") {
```
Comment (jkbradley) [inline mllib/src/test/scala/org/apache/spark/ml/feature/NormalizerSuite.scala]: Insert newline above to separate tests

**Comment by jkbradley** [comment]
Thanks for adding the test suite!  I added 2 small comments.  My only other thought is that there is some code duplication in the test suite.  If you can shorten the code by eliminating duplication, that would be great.  Other than these items, I think it's good.

**Comment by yinxusen** [comment]
I'll shorten the code. Thanks for the comments.

**Comment by jkbradley** [comment]
LGTM!  Merging into master

**Pull request closed without merging**